crates/sui-core/src/execution_cache/writeback_cache.rs

```diff
@@ -0,0 +1,1267 @@
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            ObjectEntry::Object(o) => {
+                write!(f, "ObjectEntry::Object({:?})", o.compute_object_reference())
+            }
+            ObjectEntry::Deleted => write!(f, "ObjectEntry::Deleted"),
+            ObjectEntry::Wrapped => write!(f, "ObjectEntry::Wrapped"),
+        }
+    }
+}
+
+impl From<Object> for ObjectEntry {
+    fn from(object: Object) -> Self {
+        ObjectEntry::Object(object)
+    }
+}
+
+type MarkerKey = (EpochId, ObjectID);
+
+enum CacheResult<T> {
+    /// Entry is in the cache
+    Hit(T),
+    /// Entry is not in the cache and is known to not exist
+    NegativeHit,
+    /// Entry is not in the cache and may or may not exist in the store
+    Miss,
+}
+
+/// UncommitedData stores execution outputs that are not yet written to the db. Entries in this
+/// struct can only be purged after they are committed.
+struct UncommittedData {
+    /// The object dirty set. All writes go into this table first. After we flush the data to the
+    /// db, the data is removed from this table and inserted into the object_cache.
+    ///
+    /// This table may contain both live and dead objects, since we flush both live and dead
+    /// objects to the db in order to support past object queries on fullnodes.
+    ///
+    /// Further, we only remove objects in FIFO order, which ensures that the the cached
+    /// sequence of objects has no gaps. In other words, if we have versions 4, 8, 13 of
+    /// an object, we can deduce that version 9 does not exist. This also makes child object
+    /// reads efficient. `object_cache` cannot contain a more recent version of an object than
+    /// `objects`, and neither can have any gaps. Therefore if there is any object <= the version
+    /// bound for a child read in objects, it is the correct object to return.
+    objects: DashMap<ObjectID, CachedVersionMap<ObjectEntry>>,
+
+    // Markers for received objects and deleted shared objects. This contains all of the dirty
+    // marker state, which is committed to the db at the same time as other transaction data.
+    // After markers are committed to the db we remove them from this table and insert them into
+    // marker_cache.
+    markers: DashMap<MarkerKey, CachedVersionMap<MarkerValue>>,
+
+    transaction_effects: DashMap<TransactionEffectsDigest, TransactionEffects>,
+
+    // Because TransactionEvents are not unique to the transaction that created them, we must
+    // reference count them in order to know when we can remove them from the cache. For now
+    // we track all referers explicitly, but we can use a ref count when we are confident in
+    // the correctness of the code.
+    transaction_events:
+        DashMap<TransactionEventsDigest, (BTreeSet<TransactionDigest>, TransactionEvents)>,
+
+    executed_effects_digests: DashMap<TransactionDigest, TransactionEffectsDigest>,
```

Comment:
I don't think we can do that - we need to store TransactionEffects by the effects digest for state sync. Changing that would be difficult and create a lot of problems. Otherwise, you don't have anywhere to store uncertified effects - storing them with the TransactionDigest can cause things to go badly wrong in a fork. For instance, we rely on the fact that we can insert the change epoch effects based on their digest before we have a certified checkpoint.